File: src/JPAKEEllipticCurvePOJOs/SchnorrZKP.java
package JPAKEEllipticCurvePOJOs;

import org.bouncycastle.math.ec.ECPoint;

import java.io.Serializable;
import java.math.BigInteger;
import java.nio.ByteBuffer;
import java.security.MessageDigest;
import java.security.SecureRandom;

public class SchnorrZKP implements Serializable {

    private byte[] V = null;
    private BigInteger r = null;

    public SchnorrZKP () {
        // constructor
    }

    public void generateZKP (ECPoint generator, BigInteger n, BigInteger x, ECPoint X, String userID) {

        /* Generate a random v from [1, n-1], and compute V = G*v */
        BigInteger v = org.bouncycastle.util.BigIntegers.createRandomInRange(BigInteger.ONE,
                n.subtract(BigInteger.ONE), new SecureRandom());
        ECPoint vEC = generator.multiply(v);
        V = vEC.getEncoded(false);

        BigInteger h = getSHA256(generator, vEC, X, userID); // h

        r = v.subtract(x.multiply(h)).mod(n); // r = v-x*h mod n
    }

    public byte[] getV() {
        return V;
    }

    public BigInteger getr() {
        return r;
    }
    public BigInteger getSHA256(ECPoint generator, ECPoint V, ECPoint X, String userID) {

        MessageDigest sha256 = null;

        try {
            sha256 = MessageDigest.getInstance("SHA-256");

            byte [] GBytes = generator.getEncoded(false);
            byte [] VBytes = V.getEncoded(false);
            byte [] XBytes = X.getEncoded(false);
            byte [] userIDBytes = userID.getBytes();

            // It's good practice to prepend each item with a 4-byte length
            sha256.update(ByteBuffer.allocate(4).putInt(GBytes.length).array());
            sha256.update(GBytes);

            sha256.update(ByteBuffer.allocate(4).putInt(VBytes.length).array());
            sha256.update(VBytes);

            sha256.update(ByteBuffer.allocate(4).putInt(XBytes.length).array());
            sha256.update(XBytes);

            sha256.update(ByteBuffer.allocate(4).putInt(userIDBytes.length).array());
            sha256.update(userIDBytes);

        } catch (Exception e) {
            e.printStackTrace();
        }

        return new BigInteger(sha256.digest());
    }
}

File: pythonJPAKEPlusECDemo/SchnorrZKP.py
import random 
import hashlib
class SchnorrZKP:
    '''
        This class provides the functionality to generate a 
        schnorr zero-knowledge proof using ecc.
    '''
    V = 0
    r = 0

    def __init__(self): 
        pass
    
    def generateZKP(self, cv, generator, N, x, X, userID):
        '''
            function: generateZKP
            This function creates the parameters needed for a CPZKP.
            params: @cv Curve
                    @generator Point
                    @N Point
                    @x int
                    @X Point
                    @signerID string
            @return Null
        '''
        # Generate a random v from [1, n-1], and compute V = G*v 
        v = random.randint(1, N)
        self.V = generator.mul(v);

        h = self.getSHA256(cv, generator, self.V, X, userID) # h
        
        self.r = (v - (x * h)) % N # r = v-x*h mod n

    def getV(self):
        return self.V;

    def getR(self):
        return self.r

    def getSHA256(self, cv, generator, V, X, userID):
        '''
            function: getSHA256_ec
            This function hashes multiple ec points with a 
            string using sha256.
            params: @cv curve elliptic curve
                    @generator Point ec point
                    @V Point ec point
                    @X Point ec point
                    @userID string user id
            return: int
        '''
        m = hashlib.sha256()
        try:
            GBytes = bytearray(cv.encode_point(generator));
            VBytes = bytearray(cv.encode_point(V))
            XBytes = bytearray(cv.encode_point(X));
            userIDBytes = bytearray()
            userIDBytes.extend(map(ord, str(userID)))
            # It's good practice to prepend each item with a 4-byte length
            m.update(bytearray(len(GBytes))[:4]);
            m.update(GBytes);

            m.update(bytearray(len(VBytes))[:4]);
            m.update(VBytes);

            m.update(bytearray(len(XBytes))[:4]);
            m.update(XBytes);

            m.update(bytearray(len(userIDBytes))[:4]);
            m.update(userIDBytes);
        except Exception as e:
            print(e)
            print("SHA256 failed")
        b = m.hexdigest()
        as_int = int(b, 16)
        return as_int
